jax/core.py

```diff
@@ -1397,15 +1444,23 @@ def typecheck(aval: AbstractValue, x) -> bool:
   return typecompat(aval, get_aval(x))
 
 def typecompat(aval_ref: AbstractValue, aval: AbstractValue) -> bool:
-  """Determine whether `aval` conforms to `aval_ref`"""
-  aval_ref = raise_to_shaped(aval_ref).strip_weak_type()
+  """Determine whether `aval` conforms to `aval_ref`.
+
+  Ignores weak_type and named_shape.
```

Comment:
This seems like funny behavior for `typecompat` (and `typematch` below). Is it temporary? If so, it might be good to note that, in a comment or the docstring.

Reply:
It's intentionally encoding current behavior, and after the revisions they've gone through over the past month or so none of the current plans would require us to change it.

That said, there are other reasons why we might want to hold jaxprs themselves to a stricter notion of type compatibility than the Python-facing programming model, even if doing this means extra/duplicate fixed points (i.e., the fixed points in the traceables that I added but now believe are both unnecessary and insufficient). One idea is to tighten these checks if/when we can annotate jaxprs with "names in which they're impure" (where those are the names we would want to strictly check).

Reply:
IIUC you're saying it's not temporary. But then this seems like undesirable behavior: for instance, it'd be saying the jaxpr types `f32[;i=5]`, `f32[;i=6]`, and `f32[;j=5,k=12]` are all equal. Is that right?

I think we should probably figure out how to avoid that. Ignoring named shapes temporarily in type equality could make sense as a way to land the change in pieces, but we should probably iterate more on the long-term plan. A doc would be helpful!

Reply:
Yeah, I think I agree that we want to make `typematch` stricter here in the future (and that will require things like the fixed points in control flow traceables).
 
For `typecompat`, though, we really do mean to say that a value of type `f32[;i=5]` is compatible with a variable of type `f32[;j=5,k=12]`: binding a value like that means performing a pure map over `'i'` and broadcasting over `'j'` and `'k'`. For a while, we thought that we needed two separate versions of type compatibility—one stricter version for jaxprs and the other (with this looser behavior) for tracers, but in practice it turns out that every call site for `typecompat` seems to need the looser version.

(And I agree, the changing plans here are worth writing down for the rest of the team.)

Reply:
> but in practice it turns out that every call site for `typecompat` seems to need the looser version

Well, perhaps one way forward is to have that looser version, but perhaps there are other ways too, like making the broadcasating you're talking about more explicit rather than implicit (say, in `scan`). Can we treat this as an undecided design question for now and iterate together with some concrete examples?

Reply:
> Can we treat this as an undecided design question for now and iterate together with some concrete examples?

Yes, I think so.

Broadly, one example I keep in mind is a `psum` in a loop, which we allow today. Jaxpr requires that loop bodies have type `T -> T`, but `psum` drops the formerly-implicit named axis. (This has led us before to discuss a `broadcasting_sum`.) Previously we got away with treating:

```
{ lambda a .
  let b = psum['i'] a
  in [b] }
```

as though it is `T -> T`. In some sense, this jaxpr wasn't even a function before; it doesn't make sense to ask whether it is `T -> T`. Meanwhile, the proper function that this jaxpr implied in full context wasn't `T -> T` either, but rather `T[i=N] -> T`. The explicit names are bringing this to our attention now.

A stopgap for preserving behavior like this—which is a need for now—is to keep type compatibility loose such that some of our internals continue to view this as `T -> T`. I'm actually not sure whether changing `typecompat` would directly break this loop example in terms of tests. Maybe @jekbradbury has a more directly relevant example.

Reply:
I think the semantics were consistent before because we defined psum to do a reduce-sum-broadcast, i.e. that jaxpr would be `T[i=N] -> T[i=N]`. But now that we are changing that (and have already changed it in e.g. how vmap psum works), it has implications.

But it's a good example to frame the question concretely: do we allow that jaxpr to be `T[i=N] -> T` and say it can be well-typed in a loop, or do we demand that to be used in a loop it be `T[i=N] -> T[i=N]`, at least by the time it reaches a jaxpr if not in the surface API?

Reply:
Well put!

> all of which need this looser notion of type compatibility

I'm not sure of that yet; there may be alternatives, like teaching those functions about batching over names explicitly.

> And I know this is a change of tune from what I was saying a few weeks ago

Hey man I'm changing my tune by the day. Your tune is a relative rock of stability! (Thanks @frostig and @jekbradbury for the excellent discussion so far. It is a good review for me.)

Reply:
There are certainly multiple approaches for how to implement these jaxpr eval functions, but ultimately I think all of them need to know something along the lines of "does this value inhabit one of the batching-based family of types that this jaxpr supports for this binder". On the other hand, only jaxpr typechecking would need the stricter type compatibility check (and only if we do things like fixpoint the loop body to `T[i=N] -> T[i=N]`, which I'm no longer sure we want to do—otherwise I believe jaxpr typechecking needs the same looser compatibility definition.)

Reply:
Worth noting that the current batching rule for scan (and therefore the named-shape transfer function after PR 3) would make the return value of such a scan mapped; making it unmapped could be a later optimization to the transfer function that requires the scan length to be >0 (our batching rules, unsurprisingly, aren't yet optimized much for collectives).